### PR TITLE
WDP-70

### PR DESCRIPTION
**Pull request by lukaszcwiklinski**
https://kindergarten.stxnext.pl/browse/WDP-70

### CODE REVIEW

File: src/presence_analyzer/views.py [hunk 1]
```diff
@@ -7,7 +7,8 @@
 from flask import redirect, abort
 
 from main import app
-from utils import jsonify, get_data, mean, group_by_weekday
+from utils import jsonify, get_data, mean, group_by_weekday, \
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: use () instead of \

File: src/presence_analyzer/utils.py [hunk 2]
```diff
@@ -82,6 +82,24 @@ def group_by_weekday(items):
...
+    time_start = [[] for i in range(7)]
+    time_end = [[] for i in range(7)]
+
+    for date in items:
+        start = items[date]['start']
+        end = items[date]['end']
+        time_start[date.weekday()].append(seconds_since_midnight(start))
+        time_end[date.weekday()].append(seconds_since_midnight(end))
+
+    result = [[], [], [], [], [], [], []]  # one list for every day in week.
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: [[] for i in range(7)]

File: src/presence_analyzer/views.py [hunk 3]
```diff
@@ -73,3 +74,21 @@ def presence_weekday_view(user_id):
...
+@app.route('/api/v1/presence_start_end/<int:user_id>', methods=['GET'])
+@jsonify
+def mean_time_of_start_and_end_work(user_id):
+    """
+    Returns mean time of start and end of work.
+    """
+    data = get_data()
+    if user_id not in data:
+        log.debug('User %s not found!', user_id)
+        abort(404)
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: Empty line after this one

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 4]
```diff
@@ -23,7 +23,7 @@
             padding: 0 2em;
             border: 1px solid #AAA;
             background-color: #FFF;
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: lower case

File: src/presence_analyzer/static/presence_start_end.html [hunk 5]
```diff
@@ -23,7 +23,7 @@
             padding: 0 2em;
             border: 1px solid #AAA;
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: ditto

File: src/presence_analyzer/static/presence_start_end.html [hunk 6]
```diff
@@ -79,8 +79,16 @@
         google.load("visualization", "1", {packages:["corechart", "timeline"], 'language': 'pl'});
     </script>
     <script type="text/javascript">
+        function parseInterval(value) {
+            var result = new Date(1,1,1);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: 1, 1, 1

File: src/presence_analyzer/static/presence_start_end.html [hunk 7]
```diff
@@ -79,8 +79,16 @@
...
+            result.setMilliseconds(value*1000);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: ditto

File: src/presence_analyzer/utils.py [hunk 8]
```diff
@@ -75,13 +75,31 @@ def group_by_weekday(items):
     """
     Groups presence entries by weekday.
     """
-    result = [[], [], [], [], [], [], []]  # one list for every day in week
+    result = [[] for i in range(7)]  # one list for every day in week.
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: of the week

File: src/presence_analyzer/static/presence_start_end.html [hunk 9]
```diff
@@ -96,33 +104,28 @@
...
+                            $.each(result, function(index, value) {
+                                value[1] = parseInterval(value[1]);
+                                value[2] = parseInterval(value[2]);
+                            });
+                            var data = new google.visualization.DataTable(result);
+                            data.addColumn('string', 'Weekday');
+                            data.addColumn({ type: 'datetime', id: 'Start' });
+                            data.addColumn({ type: 'datetime', id: 'End' });
+                            data.addRows(result);
+                            var options = {
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: Move to the top of the scope

File: src/presence_analyzer/static/presence_start_end.html [hunk 10]
```diff
@@ -96,33 +104,28 @@
...
+                                hAxis: {title: 'Weekday'}
+                            };
+                            var formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'});
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: ditto

File: src/presence_analyzer/static/presence_start_end.html [hunk 11]
```diff
@@ -96,33 +104,28 @@
...
-                        chart_div.show();
-                        loading.hide();
-                        var chart = new google.visualization.Timeline(chart_div[0]);
-                        chart.draw(data, options);
+                        $.getJSON("/api/v1/presence_start_end/"+selected_user, function(result) {
+                            $.each(result, function(index, value) {
+                                value[1] = parseInterval(value[1]);
+                                value[2] = parseInterval(value[2]);
+                            });
+                            var data = new google.visualization.DataTable(result);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: declare data above.

File: src/presence_analyzer/utils.py [hunk 12]
```diff
@@ -75,13 +75,31 @@ def group_by_weekday(items):
...
         end = items[date]['end']
         result[date.weekday()].append(interval(start, end))
     return result
 
+def group_by_weekday_by_start_end(items):
+    """
+    Groups presence entries by weekday start and end.
+    """
+    time_start = [[] for i in range(7)]
+    time_end = [[] for i in range(7)]
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: Just a tip for futhure:
image

File: src/presence_analyzer/utils.py [hunk 13]
```diff
@@ -75,13 +75,31 @@ def group_by_weekday(items):
...
+
+    for date in items:
+        start = items[date]['start']
+        end = items[date]['end']
+        time_start[date.weekday()].append(seconds_since_midnight(start))
+        time_end[date.weekday()].append(seconds_since_midnight(end))
+
+    result = [[] for i in range(7)]  # one list for every day in week.
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: of the

File: src/presence_analyzer/views.py [hunk 14]
```diff
@@ -7,7 +7,13 @@
 from flask import redirect, abort
 
 from main import app
-from utils import jsonify, get_data, mean, group_by_weekday
+from utils import (
+    jsonify,
+    get_data,
+    mean,
+    group_by_weekday,
+    group_by_weekday_by_start_end,
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: sort

File: src/presence_analyzer/static/presence_start_end.html [hunk 15]
```diff
@@ -105,19 +105,22 @@
                         loading.show();
                         chart_div.hide();
                         $.getJSON("/api/v1/presence_start_end/"+selected_user, function(result) {
+                            var data, formatter;
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: var x,
        y
        z;

File: src/presence_analyzer/static/presence_start_end.html [hunk 16]
```diff
@@ -105,19 +105,22 @@
...
                                 value[1] = parseInterval(value[1]);
                                 value[2] = parseInterval(value[2]);
                             });
-                            var data = new google.visualization.DataTable(result);
+
                             data.addColumn('string', 'Weekday');
                             data.addColumn({ type: 'datetime', id: 'Start' });
                             data.addColumn({ type: 'datetime', id: 'End' });
                             data.addRows(result);
                             var options = {
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: ?

File: src/presence_analyzer/static/presence_start_end.html [hunk 17]
```diff
@@ -96,33 +104,31 @@
...
-                        };
-                        var formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'});
-                        formatter.format(data, 1);
-                        formatter.format(data, 2);
-
-                        chart_div.show();
-                        loading.hide();
-                        var chart = new google.visualization.Timeline(chart_div[0]);
-                        chart.draw(data, options);
+                        $.getJSON("/api/v1/presence_start_end/"+selected_user, function(result) {
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: space around + 
image

File: src/presence_analyzer/static/presence_start_end.html [hunk 18]
```diff
@@ -96,33 +104,31 @@
...
+                            var options = {
+                                hAxis: {title: 'Weekday'}
+                            };
+
+                            formatter.format(data, 1);
+                            formatter.format(data, 2);
+
+                            chart_div.show();
+                            loading.hide();
+                            var chart = new google.visualization.Timeline(chart_div[0]);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: declare at the top

File: src/presence_analyzer/static/presence_weekday.html [hunk 19]
```diff
@@ -95,7 +95,7 @@
                     if(selected_user) {
                         loading.show();
                         chart_div.hide();
-                        $.getJSON("/api/v1/presence_weekday/"+selected_user, function(result) {
+                        $.getJSON("/api/v1/mean_time_weekday/"+selected_user, function(result) {
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_weekday.html]: ditto

File: src/presence_analyzer/utils.py [hunk 20]
```diff
@@ -75,13 +75,31 @@ def group_by_weekday(items):
...
+    Groups presence entries by weekday start and end.
+    """
+    time_start = [[] for i in range(7)]
+    time_end = [[] for i in range(7)]
+
+    for date in items:
+        start = items[date]['start']
+        end = items[date]['end']
+        time_start[date.weekday()].append(seconds_since_midnight(start))
+        time_end[date.weekday()].append(seconds_since_midnight(end))
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: Make it a dict, structure can be for example:

    
    {
            0: {
                'start': [...],
                'end': [...],
            },
            (...)
    }

And adjust rest of ur code if needed

File: src/presence_analyzer/static/presence_start_end.html [hunk 21]
```diff
@@ -79,6 +79,12 @@
         google.load("visualization", "1", {packages:["corechart", "timeline"], 'language': 'pl'});
     </script>
     <script type="text/javascript">
+        function parseInterval(value) {
+            var result = new Date(1, 1, 1);
+            result.setMilliseconds(value*1000);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: space around *

File: src/presence_analyzer/static/presence_start_end.html [hunk 22]
```diff
@@ -96,33 +102,30 @@
...
-                        chart_div.show();
-                        loading.hide();
-                        var chart = new google.visualization.Timeline(chart_div[0]);
-                        chart.draw(data, options);
+                        $.getJSON("/api/v1/presence_start_end/" + selected_user, function(result) {
+                            var data = new google.visualization.DataTable(result),
+                                chart = new google.visualization.Timeline(chart_div[0]),
+                                formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'}),
+                                options = {
+                                hAxis: {title: 'Weekday'}
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: semicolon

File: src/presence_analyzer/static/presence_start_end.html [hunk 23]
```diff
@@ -79,14 +79,23 @@
...
             $(document).ready(function(){
+                console.log('1');
                 var loading = $('#loading');
+                console.log('2');
                 $.getJSON("/api/v1/users", function(result) {
                     var dropdown = $("#user_id");
                     $.each(result, function(item) {
                         dropdown.append($("<option />").val(this.user_id).text(this.name));
                     });
+                    console.log('3');
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: Remove

File: src/presence_analyzer/utils.py [hunk 24]
```diff
@@ -75,13 +75,30 @@ def group_by_weekday(items):
...
+    for date in items:
+        start = items[date]['start']
+        end = items[date]['end']
+        time[date.weekday()]['start'].append(seconds_since_midnight(start))
+        time[date.weekday()]['end'].append(seconds_since_midnight(end))
+
+    result = [[] for i in range(7)] # one list for every day of the week.
+
+    for idx, weekday in enumerate(time):
+        result[idx] = [mean(weekday['start']), mean(weekday['end'])]
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: U don't rly need this loop. just results = [{'start': [], 'end': []} for i in range(7)] and fill it directly in upper loop. You should achieve the same results.

File: src/presence_analyzer/tests.py [hunk 25]
```diff
@@ -66,6 +66,14 @@ def test_presence_by_weekday_header(self):
         self.assertEqual(resp.status_code, 200)
         self.assertIn('Presence by weekday', resp.data)
 
+    def test_presence_by_start_end(self):
+        """
+        Test presence by start and end of work.
+        """
+        resp = self.client.get('/static/presence_start_end.html')
+        self.assertEqual(resp.status_code, 200)
+        self.assertIn('Presence start-end weekday', resp.data)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: data is not important?

Comment (lukaszcwiklinski) [inline src/presence_analyzer/tests.py]: I check data for start-end on line 246 here https://github.com/stxnext-kindergarten/stxnext-kindergarten-lcwiklinski/pull/6/commits/1c6d062f04ac0c53a42b39efb179b15577c34a93#diff-06804ee992f7c6b264fc1882858d9b78

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 26]
```diff
@@ -102,25 +102,26 @@
                     if(selected_user) {
                         loading.show();
                         chart_div.hide();
-                        $.getJSON("/api/v1/mean_time_weekday/"+selected_user, function(result) {
+                        $.getJSON("/api/v1/mean_time_weekday/" + selected_user, function(result) {
+                            var data = new google.visualization.DataTable(),
+                                chart = new google.visualization.ColumnChart(chart_div[0]),
+                                formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'}),
+                                options = {
+                                hAxis: {title: 'Weekday'}
```
Comment (dpolcyn) [inline src/presence_analyzer/static/mean_time_weekday.html]: style

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 27]
```diff
@@ -21,9 +21,9 @@
         }
         #content {
             padding: 0 2em;
-            border: 1px solid #AAA;
-            background-color: #FFF;
-            
+            border: 1px solid #aaa;
+            background-color: #fff;
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: remove

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 28]
```diff
@@ -102,25 +102,31 @@
...
+                        function(result) {
+                            var data = new google.visualization.DataTable(),
+                                chart = new google.visualization.ColumnChart(chart_div[0]),
+                                formatter = new google.visualization.DateFormat({
+                                    pattern: 'HH:mm:ss',
+                                }),
+                                options = {
+                                    hAxis: {
+                                        title: 'Weekday',
+                                    },
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: remove ,

File: src/presence_analyzer/static/presence_start_end.html [hunk 29]
```diff
@@ -21,9 +21,9 @@
         }
         #content {
             padding: 0 2em;
-            border: 1px solid #AAA;
-            background-color: #FFF;
-            
+            border: 1px solid #aaa;
+            background-color: #fff;
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: :policeman:

File: src/presence_analyzer/static/presence_start_end.html [hunk 30]
```diff
@@ -96,33 +102,35 @@
...
+                                    },
+                                };
+
+                            $.each(result, function(index, value) {
+                                value[1] = parseInterval(value[1]);
+                                value[2] = parseInterval(value[2]);
+                            });
 
+                            data.addColumn('string', 'Weekday');
+                            data.addColumn({ type: 'datetime', id: 'Start' });
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: data.addColumn({type: 'datetime', id: 'Start'});

File: src/presence_analyzer/static/presence_start_end.html [hunk 31]
```diff
@@ -96,33 +102,35 @@
...
+                            data.addColumn({ type: 'datetime', id: 'End' });
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: :policeman:

File: .gitignore [hunk 32]
```diff
@@ -1,7 +1,6 @@
 *.py[cod]
 *.idea/
 
-
```
Comment (dpolcyn) [inline .gitignore]: ??????????????????????

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 33]
```diff
@@ -7,24 +7,21 @@
...
+            background-color: #eee;
         }
         #main {
             width: 800px;
             margin: 2em auto;
             padding: 0 2em;
         }
         #content {
-            padding: 0 2em;
-            border: 1px solid #AAA;
```
Comment (dpolcyn) [inline src/presence_analyzer/static/mean_time_weekday.html]: ????????????????????????????????????

File: src/presence_analyzer/static/presence_start_end.html [hunk 34]
```diff
@@ -79,6 +76,12 @@
         google.load("visualization", "1", {packages:["corechart", "timeline"], 'language': 'pl'});
     </script>
     <script type="text/javascript">
+        function parseInterval(value) {
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: move to utils.js and use it istead of copy it everywhere

File: src/presence_analyzer/static/css/style.css [hunk 35]
```diff
@@ -0,0 +1,58 @@
+body {
+    font-family: latoregular, Helvetica, Arial;
+    background-color: #eee;
+}
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/css/style.css]: enter after each }

File: src/presence_analyzer/static/css/style.css [hunk 36]
```diff
@@ -0,0 +1,58 @@
...
+    border-color: black;
+}
+
+#header #selected a {
+    position: relative;
+    top: 1px;
+    background: white;
+    color: black;
+    font-weight: bold;
+}
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/css/style.css]: Configure ur editor

File: src/presence_analyzer/static/js/utils.js [hunk 37]
```diff
@@ -0,0 +1,7 @@
+
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/js/utils.js]: ???

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 38]
```diff
@@ -102,25 +36,31 @@
                     if(selected_user) {
                         loading.show();
                         chart_div.hide();
-                        $.getJSON("/api/v1/mean_time_weekday/"+selected_user, function(result) {
+                        $.getJSON("/api/v1/mean_time_weekday/" + selected_user,
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: $.getJSON("/api/v1/mean_time_weekday/" + selected_user, function (result) {

File: src/presence_analyzer/static/presence_start_end.html [hunk 39]
```diff
@@ -7,78 +7,18 @@
...
+    <link href="/static/css/style.css" media="all" rel="stylesheet" type="text/css" />
 
     <script src="/static/js/jquery.min.js"></script>
+    <script src="/static/js/utils.js"></script>
     <script type="text/javascript" src="https://www.google.com/jsapi"></script>
     <script type="text/javascript">
         google.load("visualization", "1", {packages:["corechart", "timeline"], 'language': 'pl'});
     </script>
     <script type="text/javascript">
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: remove

File: src/presence_analyzer/static/presence_weekday.html [hunk 40]
```diff
@@ -95,12 +33,12 @@
                     if(selected_user) {
                         loading.show();
                         chart_div.hide();
-                        $.getJSON("/api/v1/presence_weekday/"+selected_user, function(result) {
-                            var data = google.visualization.arrayToDataTable(result);
-                            var options = {};
+                        $.getJSON("/api/v1/mean_time_weekday/" + selected_user, function(result) {
+                            var data = google.visualization.arrayToDataTable(result),
+                                chart = new google.visualization.PieChart(chart_div[0]),
+                                options = {};
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_weekday.html]: enter for readability.

File: src/presence_analyzer/tests.py [hunk 41]
```diff
@@ -213,6 +239,39 @@ def test_group_by_weekday(self):
...
+                {
+                    0: {'start': [], 'end': []},
+                    1: {'start': [0], 'end': [0]},
+                    2: {'start': [33592], 'end': [58057]},
+                    3: {'start': [38926], 'end': [62631]},
+                    4: {'start': [], 'end': []},
+                    5: {'start': [], 'end': []},
+                    6: {'start': [], 'end': []},
+                }
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: ?
style..

File: src/presence_analyzer/tests.py [hunk 42]
```diff
@@ -213,6 +239,39 @@ def test_group_by_weekday(self):
...
+                        'end': datetime.time(0, 0, 0),
+                    },
+                    datetime.date(2013, 9, 12): {
+                        'start': datetime.time(10, 48, 46),
+                        'end': datetime.time(17, 23, 51),
+                    },
+                    datetime.date(2013, 9, 11): {
+                        'start': datetime.time(9, 19, 52),
+                        'end': datetime.time(16, 7, 37),
+                    },
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: remove ,

**Pull request merged**